libcontainer/nsenter/nsexec.c

```diff
@@ -56,166 +54,248 @@ int setns(int fd, int nstype)
 #endif
 #endif
 
-static int clone_parent(jmp_buf * env) __attribute__ ((noinline));
-static int clone_parent(jmp_buf * env)
+static int clone_parent(jmp_buf * env, int flags) __attribute__ ((noinline));
+static int clone_parent(jmp_buf * env, int flags)
 {
 	struct clone_arg ca;
 	int child;
 
 	ca.env = env;
-	child = clone(child_func, ca.stack_ptr, CLONE_PARENT | SIGCHLD, &ca);
-
+	child = clone(child_func, ca.stack_ptr, CLONE_PARENT | SIGCHLD | flags, &ca);
 	return child;
 }
 
+// get init pipe from the parent. It's used to read bootstrap data, and to
+// write pid to after nsexec finishes setting up the environment.
+static int get_init_pipe()
+{
+	char buf[PATH_MAX], *initpipe;
+	int pipenum = -1;
+
+	initpipe = getenv("_LIBCONTAINER_INITPIPE");
+	if (initpipe == NULL) {
+		return -1;
+	}
+
+	pipenum = atoi(initpipe);
+	snprintf(buf, sizeof(buf), "%d", pipenum);
+	if (strcmp(initpipe, buf)) {
+		pr_perror("Unable to parse _LIBCONTAINER_INITPIPE");
+		exit(1);
+	}
+
+	return pipenum;
+}
+
+// num_namespaces returns the number of additional namespaces to setns. The
+// argument is a comma-separated string of namespace paths.
+static int num_namespaces(char *nspaths)
+{
+	int size = 0, i = 0;
+
+	for (i = 0; nspaths[i]; i++) {
+		if (nspaths[i] == ',') {
+			size += 1;
+		}
+	}
+
+	return size + 1;
+}
+
 static uint32_t readint32(char *buf)
 {
-	return *(uint32_t *) buf;
+	return *(uint32_t *)buf;
+}
+
+static uint8_t readint8(char *buf)
+{
+	return *(uint8_t *)buf;
+}
+
+static void writedata(int fd, char *data, int start, int len)
+{
+	int written = 0;
+	while (written < len) {
+		size_t nbyte, i;
+		if ((len - written) < 1024) {
+			nbyte = len - written;
+		} else {
+			nbyte = 1024;
+		}
+		i = write(fd, data + start + written, nbyte);
+		if (i == -1) {
+			pr_perror("failed to write data to %d", fd);
+			exit(1);
+		}
+		written += i;
+	}
 }
 
 // list of known message types we want to send to bootstrap program
 // These are defined in libcontainer/message_linux.go
-#define INIT_MSG 62000
-#define PID_ATTR 27281
-#define CONSOLE_PATH_ATTR 27282
+#define INIT_MSG			62000
+#define CLONE_FLAGS_ATTR	27281
+#define CONSOLE_PATH_ATTR	27282
+#define NS_PATHS_ATTR		27283
+#define UIDMAP_ATTR			27284
+#define GIDMAP_ATTR			27285
+#define SETGROUP_ATTR		27286
```

Comment:
i think there are some misalignments here when you rebased. The same for a few other instances below :D